fl4health/losses/deep_mmd_loss.py

```diff
@@ -0,0 +1,224 @@
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        """Forward the LeNet."""
+        fealant = self.latent(input)
+        return fealant
+
+
+class DeepMmdLoss(torch.nn.Module):
+    def __init__(
+        self,
+        device: torch.device,
+        input_size: int,
+        lr: float = 0.001,
+        hidden_size: int = 10,
+        output_size: int = 50,
+        layer_name: Optional[str] = None,
+        training: bool = True,
+    ) -> None:
+        super().__init__()
+        self.device = device
+        self.lr = lr
+        self.layer_name = layer_name
+        self.training = training
+
+        # Initialize the model
+        self.featurizer = ModelLatentF(input_size, hidden_size, output_size).to(self.device)
+
+        # Initialize parameters
+        self.epsilonOPT: torch.Tensor = torch.log(torch.from_numpy(np.random.rand(1) * 10 ** (-10)).to(self.device))
+        self.epsilonOPT.requires_grad = self.training
+        self.sigmaOPT: torch.Tensor = torch.sqrt(torch.tensor(2 * 32 * 32, dtype=torch.float).to(self.device))
+        self.sigmaOPT.requires_grad = self.training
+        self.sigma0OPT: torch.Tensor = torch.sqrt(torch.tensor(0.005, dtype=torch.float).to(self.device))
+        self.sigma0OPT.requires_grad = self.training
+
+        # Initialize optimizers
+        self.optimizer_F = torch.optim.Adam(
+            list(self.featurizer.parameters()) + [self.epsilonOPT] + [self.sigmaOPT] + [self.sigma0OPT], lr=self.lr
+        )
+
+    def Pdist2(self, x: torch.Tensor, y: Optional[torch.Tensor]) -> torch.Tensor:
+        """compute the paired distance between x and y."""
+        x_norm = (x**2).sum(1).view(-1, 1)
+        if y is not None:
+            y_norm = (y**2).sum(1).view(1, -1)
+        else:
+            y = x
+            y_norm = x_norm.view(1, -1)
+        Pdist = x_norm + y_norm - 2.0 * torch.mm(x, torch.transpose(y, 0, 1))
+        Pdist[Pdist < 0] = 0
+        return Pdist
+
+    def h1_mean_var_gram(
+        self, Kx: torch.Tensor, Ky: torch.Tensor, Kxy: torch.Tensor, is_var_computed: bool, use_1sample_U: bool = True
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], torch.Tensor]:
+        """compute value of MMD and std of MMD using kernel matrix."""
+        Kxxy = torch.cat((Kx, Kxy), 1)
+        Kyxy = torch.cat((Kxy.transpose(0, 1), Ky), 1)
+        Kxyxy = torch.cat((Kxxy, Kyxy), 0)
+        nx = Kx.shape[0]
+        ny = Ky.shape[0]
+        is_unbiased = True
+        if is_unbiased:
+            xx = torch.div((torch.sum(Kx) - torch.sum(torch.diag(Kx))), (nx * (nx - 1)))
+            yy = torch.div((torch.sum(Ky) - torch.sum(torch.diag(Ky))), (ny * (ny - 1)))
+            # one-sample U-statistic.
+            if use_1sample_U:
+                xy = torch.div((torch.sum(Kxy) - torch.sum(torch.diag(Kxy))), (nx * (ny - 1)))
+            else:
+                xy = torch.div(torch.sum(Kxy), (nx * ny))
+            mmd2 = xx - 2 * xy + yy
+        else:
```

Comment:
It's been a while since I left this comment but I think we are computing the very specific notation above. I think it's worth being precise (if I was current in the measure being computed), that is \hat{\text{MMD}}_b^2 in your comment.